packages/react-devtools-shared/src/backend/console.js

```diff
@@ -251,17 +254,60 @@ export function patch({
                   if (componentStack !== '') {
-                    if (isStrictModeOverride(args)) {
-                      if (__IS_FIREFOX__) {
-                        args[0] = `${args[0]} %s`;
-                        args.push(componentStack);
-                      } else {
-                        args[0] =
-                          ANSI_STYLE_DIMMING_TEMPLATE_WITH_COMPONENT_STACK;
-                        args.push(componentStack);
+                    // Create a fake Error so that when we print it we get native source maps. Every
+                    // browser will print the .stack property of the error and then parse it back for source
+                    // mapping. Rather than print the internal slot. So it doesn't matter that the internal
+                    // slot doesn't line up.
+                    const fakeError = new Error('');
+                    // In Chromium, only the stack property is printed but in Firefox the <name>:<message>
+                    // gets printed so to make the colon make sense, we name it so we print Component Stack:
+                    // and similarly Safari leave an expandable slot.
+                    fakeError.name = 'Component Stack'; // This gets printed
+                    // In Chromium, the stack property needs to start with ^[\w.]*Error\b to trigger stack
+                    // formatting. Otherwise it is left alone. So we prefix it. Otherwise we just override it
+                    // to our own stack.
+                    fakeError.stack =
+                      __IS_CHROME__ || __IS_EDGE__
+                        ? 'Error Component Stack:' + componentStack
+                        : componentStack;
+                    if (alreadyHasComponentStack) {
+                      // Only modify the component stack if it matches what we would've added anyway.
+                      // Otherwise we assume it was a non-React stack.
+                      if (
+                        areStackTracesEqual(
+                          args[args.length - 1],
+                          componentStack,
+                        )
+                      ) {
+                        args[args.length - 1] = fakeError;
+                        if (isStrictModeOverride(args)) {
+                          if (__IS_FIREFOX__) {
+                            args[0] = `${args[0]} %o`;
+                          } else {
+                            args[0] =
+                              ANSI_STYLE_DIMMING_TEMPLATE_WITH_COMPONENT_STACK;
+                          }
+                        } else {
+                          const firstArg = args[0];
+                          if (
+                            args.length > 1 &&
+                            typeof firstArg === 'string' &&
+                            firstArg.endsWith('%s')
+                          ) {
+                            args[0] = firstArg.slice(0, firstArg.length - 2); // Strip the %s param
+                          }
```

Comment:
Is this the logic for backwards-compatibility with previous versions of React? To filter out `%s` at the end of the message.

If so, why is it gated with `!isStrictModeOverride(...)`? Shouldn't it be independent of when error was emitted?

Reply:
If we're in the strict mode override we end up converting all parameterized arguments into a flat string. I'm not sure we really need to do that. E.g. for badging I tried to avoid doing that. One benefit of doing this is that we are guaranteed to be able to format the styling of all the content where as if you print them as object it might not listen to our color override.

So the formatting no longer has a trailing `%s` after it has been rewritten.

I think my intension was to just leave it as a string so that the color formatting consistently can override it. Otherwise, we'd need to use a `ANSI_STYLE_DIMMING_TEMPLATE_WITH_COMPONENT_STACK` with an `%o` for the error stack.

However I messed up some refactoring here so it's inconsistent. I'll fix that.

Reply:
> If we're in the strict mode override we end up converting all parameterized arguments into a flat string.

Oh, true. Well, we can't really guarantee that it is going to be a single flat string, because we omit other objects and user-space console stylings and just bail out.

Reply:
I'll just leave it intact in that case since it's non-essential and sometimes hidden anyway. It'll also go away with owner stacks since we'll never append then.